dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/ThreadCreation.java

```diff
@@ -334,25 +334,60 @@ private ThreadData createLLVMThreadFromFunction(Function function, int tid, Thre
         return new ThreadData(thread, null);
     }
 
-    private void replaceGlobalsByThreadLocals(Memory memory, Thread thread) {
+
+    private void replaceThreadLocalsWithStackalloc(Memory memory, Thread thread) {
+        final TypeFactory types = TypeFactory.getInstance();
+        final ExpressionFactory exprs = ExpressionFactory.getInstance();
+
+        // Translate thread-local memory object to local stack allocation
+        Map<MemoryObject, Register> toLocalRegister = new HashMap<>();
+        for (MemoryObject memoryObject : memory.getObjects()) {
+            if (!memoryObject.isThreadLocal()) {
+                continue;
+            }
+            Preconditions.checkState(memoryObject.hasKnownSize());
+
+            // Compute type of memory object based on initial values
+            final List<Type> contentTypes = new ArrayList<>();
+            final List<Integer> offsets = new ArrayList<>();
+            for (Integer initOffset : memoryObject.getInitializedFields()) {
+                contentTypes.add(memoryObject.getInitialValue(initOffset).getType());
+                offsets.add(initOffset);
+            }
+            final Type memoryType = types.getAggregateType(contentTypes, offsets);
+
+            // Allocate single object of memory type
```

Comment:
This suggests we cannot have thread local arrays, which is think it is not accurate

Reply:
Well, we currently lose content-type information during parsing, so I cannot use the original type here.
An array `__thread int array[4];` will get treated like `__thread struct {int a; int b; int c; int d} array;` Similarly, nested structs will get flattened into a single struct. I think it doesn't matter, because they all have the same memory layout.

In short, we support all types here because we only rely care about the memory layout.